### PR TITLE
Add typescript typings

### PR DESCRIPTION
**Pull request by elliotblackburn**
We were using this library with typescript so developed typings across the library, we thought it might be useful to the project to enable TS users to use the project with less friction.

Let me know if you've got any questions or would like any changes made while considering these changes.

### CODE REVIEW

**Comment by joews** [comment]
This is great, thank you!

Is there something we can put in the CI setup to verify that the typings are at least syntactically correct?

**Comment by elliotblackburn** [comment]
I had a look into this and it looks like [Microsoft/dtslint](https://github.com/Microsoft/dtslint) is what we'd want, however i'm having some problems getting it to run properly right now. They have no build indicator but I have a feeling it might be failing right now due to the errors I'm getting.

I'll take another crack this evening and see if we can integrate it into the build steps.

**Comment by joews** [comment]
The Travis hook isn't working right now. The CI build is OK, but the Github intregration isn't firing 🤔

**Comment by elliotblackburn** [comment]
It's actually dtslint locally which is causing me issues, I think they may have a broken build on npm. I'll check again today.

**Comment by elliotblackburn** [comment]
@joews I've added in dtslint and managed to get it working, it's caused a few style changes to the index.d.ts but none of the content has actually changed.

I've added this into `npm test` after the mocha run, this will test that the type file is syntactically correct. It seems the travis build still isn't firing though as you mentioned before.

**Comment by joews** [comment]
I think the Travis status checks are broken by my org's oauth settings - i'll sort that out later.

Unfortunately `dtslint` only works with Node 6+ (we support back to 0.10 at the moment), so the build is failing now: https://travis-ci.org/ripjar/passport-trusted-header/jobs/252818561

Maybe we can do something like this with Travis' `script`? https://stackoverflow.com/a/28653772/2806996

If that doesn't work, or it's too complex, we can drop it from CI.

**Comment by elliotblackburn** [comment]
I've added in a travis script as you suggested which runs `npm test` (with linting) for node 6, otherwise it defaults down to `npm run test-no-lint`. I did it this way around because I assume you'd want `npm test` to run with maximum test coverage.

Just need to see if Travis blows ups with this script now. Would you like the names changed at all, this will need modifying if you decide to test against node 7.

**Comment by elliotblackburn** [comment]
Green build from that new addition

File: .travis.yml [hunk 1]
```diff
@@ -4,3 +4,11 @@ node_js:
   - "4"
   - "0.12"
   - "0.10"
+
+script:
+  nv=$(node -v);
+  if [[ $nv == v6* ]] ; then
```
Comment (joews) [inline .travis.yml]: As you pointed out, this is quite fragile. Can you use [check-node-version](https://www.npmjs.com/package/check-node-version) to check for Node >= 6?

File: package.json [hunk 2]
```diff
@@ -3,8 +3,10 @@
   "version": "1.1.0",
   "description": "Passport.js strategy for trusted HTTP headers",
   "main": "src",
+  "typings": "types",
   "scripts": {
-    "test": "mocha"
+    "test": "mocha && dtslint types",
```
Comment (joews) [inline package.json]: I would prefer to keep `dtslint` and `mocha` separate so `npm test` only runs the tests (for a quicker turnaround in the common case of changing the logic).

Maybe `test` and `lint-types` as separate tasks, where the Travis script calls both in the "modern Node" branch?

File: types/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "lib": ["es6"],
+    "noImplicitAny": true,
```
Comment (joews) [inline types/tsconfig.json:5]: 👍  for all the types 😄

**Review by joews — CHANGES_REQUESTED** [review]
This looks great, thanks. A little tweak to the Travis config and I'll merge.

**Comment by elliotblackburn** [comment]
@joews I've been playing around with getting check-node-version working and having a few issues. Wondered if you had any ideas? The build is passing but it doesn't seem to actually be firing up the npm task `lint-types`.

I've pulled it into a shell file and and tested the statements and they run just fine but on travis it seems to be skipping over them.

A side note: I had to add `|| true` onto the command to do the check. This was because check-node-version returns status code 1 if the check fails which causes the travis build to fail instantly.

Any thoughts would be great, then we can get this squashed and merged in.

**Pull request closed without merging**